README.md
# Backend.AI Agent

The Backend.AI Agent is a small daemon that does:

* Reports the status and available resource slots of a worker to the manager
* Routes code execution requests to the designated kernel container
* Manages the lifecycle of kernel containers (create/monitor/destroy them)

## Package Structure

* `ai.backend`
  - `agent`: The agent package
    - `docker`: A docker-based backend implementation for the kernel lifecycle interface.
    - `server`: The agent daemon which communicates with the manager and the Docker daemon
    - `watcher`: A side-by-side daemon which provides a separate HTTP endpoint for accessing the status
      information of the agent daemon and manipulation of the agent's systemd service
  - `helpers`: A utility package that is available as `ai.backend.helpers` *inside* Python-based containers
  - `kernel`: Language-specific runtimes (mostly ipykernel client adaptor) which run *inside* containers
  - `runner`: Auxiliary components (usually self-contained binaries) mounted *inside* contaienrs


## Installation

Please visit [the installation guides](https://github.com/lablup/backend.ai/wiki).


### Kernel/system configuration

#### Recommended kernel parameters in the bootloader (e.g., Grub):

```
cgroup_enable=memory swapaccount=1
```

#### Recommended resource limits:

**`/etc/security/limits.conf`**
```
root hard nofile 512000
root soft nofile 512000
root hard nproc 65536
root soft nproc 65536
user hard nofile 512000
user soft nofile 512000
user hard nproc 65536
user soft nproc 65536
```

**sysctl**
```
fs.file-max=2048000
fs.inotify.max_user_watches=524288
net.core.somaxconn=1024
net.ipv4.tcp_max_syn_backlog=1024
net.ipv4.tcp_slow_start_after_idle=0
net.ipv4.tcp_fin_timeout=10
net.ipv4.tcp_window_scaling=1
net.ipv4.tcp_tw_reuse=1
net.ipv4.tcp_early_retrans=1
net.ipv4.ip_local_port_range="40000 65000"
net.core.rmem_max=16777216
net.core.wmem_max=16777216
net.ipv4.tcp_rmem=4096 12582912 16777216
net.ipv4.tcp_wmem=4096 12582912 16777216
net.netfilter.nf_conntrack_max=10485760
net.netfilter.nf_conntrack_tcp_timeout_established=432000
net.netfilter.nf_conntrack_tcp_timeout_close_wait=10
net.netfilter.nf_conntrack_tcp_timeout_fin_wait=10
net.netfilter.nf_conntrack_tcp_timeout_time_wait=10
```

The `ip_local_port_range` should not overlap with the container port range pool
(default: 30000 to 31000).


### For development

#### Prerequisites

* `libsnappy-dev` or `snappy-devel` system package depending on your distro
* Python 3.6 or higher with [pyenv](https://github.com/pyenv/pyenv)
and [pyenv-virtualenv](https://github.com/pyenv/pyenv-virtualenv) (optional but recommneded)
* Docker 18.03 or later with docker-compose (18.09 or later is recommended)

First, you need **a working manager installation**.
For the detailed instructions on installing the manager, please refer
[the manager's README](https://github.com/lablup/backend.ai-manager/blob/master/README.md)
and come back here again.

#### Preparing working copy

Install and activate [`git-lfs`](https://git-lfs.github.com/) to work with pre-built binaries in
`src/ai/backend/runner`.

```console
$ git lfs install
```

Next, prepare the source clone of the agent and install from it as follows.
`pyenv` is just a recommendation; you may use other virtualenv management tools.

```console
$ git clone https://github.com/lablup/backend.ai-agent agent
$ cd agent
$ pyenv virtualenv venv-agent
$ pyenv local venv-agent
$ pip install -U pip setuptools
$ pip install -U -r requirements/dev.txt
```

### Linting

We use `flake8` and `mypy` to statically check our code styles and type consistency.
Enable those linters in your favorite IDE or editor.

### Halfstack (single-node development & testing)

With the halfstack, you can run the agent simply.
Note that you need a working manager running with the halfstack already!

#### Recommended directory structure

* `backend.ai-dev`
  - `manager` (git clone from [the manager repo](https://github.com/lablup/backend.ai-manager))
  - `agent` (git clone from here)
  - `common` (git clone from [the common repo](https://github.com/lablup/backend.ai-common))

Install `backend.ai-common` as an editable package in the agent (and the manager) virtualenvs
to keep the codebase up-to-date.

```console
$ cd agent
$ pip install -U -e ../common
```

#### Steps

```console
$ mkdir -p "./scratches"
$ cp config/halfstack.toml ./agent.toml
```

Then, run it (for debugging, append a `--debug` flag):

```console
$ python -m ai.backend.agent.server
```

To run the agent-watcher:

```console
$ python -m ai.backend.agent.watcher
```

The watcher shares the same configuration TOML file with the agent.
Note that the watcher is only meaningful if the agent is installed as a systemd service
named `backendai-agent.service`.

To run tests:

```console
$ python -m flake8 src tests
$ python -m pytest -m 'not integration' tests
```


## Deployment

### Configuration

Put a TOML-formatted agent configuration (see the sample in `config/sample.toml`)
in one of the following locations:

 * `agent.toml` (current working directory)
 * `~/.config/backend.ai/agent.toml` (user-config directory)
 * `/etc/backend.ai/agent.toml` (system-config directory)

Only the first found one is used by the daemon.

The agent reads most other configurations from the etcd v3 server where the cluster
administrator or the Backend.AI manager stores all the necessary settings.

The etcd address and namespace must match with the manager to make the agent
paired and activated.
By specifying distinguished namespaces, you may share a single etcd cluster with multiple
separate Backend.AI clusters.

By default the agent uses `/var/cache/scratches` directory for making temporary
home directories used by kernel containers (the `/home/work` volume mounted in
containers).  Note that the directory must exist in prior and the agent-running
user must have ownership of it.  You can change the location by
`scratch-root` option in `agent.toml`.

### Running from a command line

The minimal command to execute:

```sh
python -m ai.backend.agent.server
python -m ai.backend.agent.watcher
```

For more arguments and options, run the command with `--help` option.

### Example config for systemd

`/etc/systemd/system/backendai-agent.service`:

```dosini
[Unit]
Description=Backend.AI Agent
Requires=docker.service
After=network.target remote-fs.target docker.service

[Service]
Type=simple
User=root
Group=root
Environment=HOME=/home/user
ExecStart=/home/user/backend.ai/agent/run-agent.sh
WorkingDirectory=/home/user/backend.ai/agent
KillMode=process
KillSignal=SIGTERM
PrivateTmp=false
Restart=on-failure
RestartSec=5

[Install]
WantedBy=multi-user.target
```

`/home/user/agent/run-agent.sh`:

```sh
#! /bin/sh
if [ -z "$PYENV_ROOT" ]; then
  export PYENV_ROOT="$HOME/.pyenv"
  export PATH="$PYENV_ROOT/bin:$PATH"
fi
eval "$(pyenv init -)"
eval "$(pyenv virtualenv-init -)"

cd /home/user/backend.ai/agent
if [ "$#" -eq 0 ]; then
  exec python -m ai.backend.agent.server
else
  exec "$@"
fi
```

### Networking

The manager and agent should run in the same local network or different
networks reachable via VPNs, whereas the manager's API service must be exposed to
the public network or another private network that users have access to.

The manager must be able to access TCP ports 6001, 6009, and 30000 to 31000 of the agents in default
configurations.  You can of course change those port numbers and ranges in the configuration.

| Manager-to-Agent TCP Ports | Usage |
|:--------------------------:|-------|
| 6001                       | ZeroMQ-based RPC calls from managers to agents |
| 6009                       | HTTP watcher API |
| 30000-31000                | Port pool for in-container services |

The operation of agent itself does not require both incoming/outgoing access to
the public Internet, but if the user's computation programs need the Internet, the docker containers
should be able to access the public Internet (maybe via some corporate firewalls).

| Agent-to-X TCP Ports     | Usage |
|:------------------------:|-------|
| manager:5002             | ZeroMQ-based event push from agents to the manager |
| etcd:2379                | etcd API access |
| redis:6379               | Redis API access |
| docker-registry:{80,443} | HTTP watcher API |
| (Other hosts)            | Depending on user program requirements |


LICENSES
--------

[GNU Lesser General Public License](https://github.com/lablup/backend.ai-agent/blob/master/LICENSE)
[Dependencies](https://github.com/lablup/backend.ai-manager/blob/agent/DEPENDENCIES.md)


HISTORY.md
Changes
=======

19.09.26 (2020-05-28)
---------------------

### Fixes
* Prevent executing startup command multiple times for batch session. ([#217](https://github.com/lablup/backend.ai-agent/issues/217))
* Fix the duplicate Docker event issue due to a misuse of the aidoocker API ([#220](https://github.com/lablup/backend.ai-agent/issues/220))
* Fix blocking of subsequent lifecycle event processing by a hanging-up handler ([#221](https://github.com/lablup/backend.ai-agent/issues/221))


19.09.25 (2020-05-20)
---------------------

### Fixes
* Improve stability under heavily loaded scenarios ([#214](https://github.com/lablup/backend.ai-agent/issues/214))
  - Skip lifecycle sync for already terminating kernels to reduce excessive Docker Engine overheads with a many number of being-terminated kernels
  - Remove timeout for container termination for both self-terminated and user-requested cases
  - Increase timeout for container termination to 60 seconds during restarting kernels, by observing deletion latencies under heavy load tests


19.09.24 (2020-05-15)
---------------------

### Fixes
* Fix lifecycle-related code errors when handling results of batch-mode tasks ([#210](https://github.com/lablup/backend.ai-agent/issues/210))
* Fix instability caused by stat-synchronizer processes under heavy loads by collecting statistics periodically only ([#212](https://github.com/lablup/backend.ai-agent/issues/212))
* Apply batching when producing "kernel_stat_sync" events to reduce manager loads and increase timeout for caching stats in Redis from 30 seconds to 2 minutes ([#213](https://github.com/lablup/backend.ai-agent/issues/213))


19.09.22 (2020-04-30)
---------------------

### Fixes
* Fortify exception handling in container lifecycle processor ([#207](https://github.com/lablup/backend.ai-agent/issues/207))

19.09.22 (2020-04-30)
---------------------

### Fixes
* Check closing of zmq sockets in code runner to avoid writing on invalid/closed sockets and indefinite waiting for the krunner execution results ([#205](https://github.com/lablup/backend.ai-agent/issues/205))
* Introduce "already-terminated" reason for `kernel_terminated` event ([#206](https://github.com/lablup/backend.ai-agent/issues/206))

19.09.21 (2020-04-30)
---------------------

### Fixes
* Ensure unhandled exceptions in aiozmq.rpc handlers to be always msgpack-serializable to keep the agent daemon logs clean ([#199](https://github.com/lablup/backend.ai-agent/issues/199))
* Add the force option when calling Docker's container deletion API to avoid rare "stop before removal" errors even when we try deletion after receiving the termination event ([#200](https://github.com/lablup/backend.ai-agent/issues/200))
* Keep the docker event processor running regardless of unexpected exceptions in the middle ([#202](https://github.com/lablup/backend.ai-agent/issues/202))
* Use destroy() of ZeroMQ context objects instead of term() to stabilize container removals ([#203](https://github.com/lablup/backend.ai-agent/issues/203))
* Delete auto-created/temporary volumes with when deleting containers ([#204](https://github.com/lablup/backend.ai-agent/issues/204))

### Miscellaneous
* Adopt towncrier for changelog management ([#201](https://github.com/lablup/backend.ai-agent/issues/201))

19.09.20 (2020-04-27)
---------------------

* FEATURE: Place bootstrap script upon session creation (#174)
* FIX: Update the restricted/reserved list of service port numbers, allowing privileged TCP ports
  to be used by the kernel image authors. (#195)

19.09.19 (2020-03-24)
---------------------

* HOTFIX: Regression in a majority of kernels due to a missing `self.loop` initialization.

19.09.18 (2020-03-21)
---------------------

* IMPROVE: Support code-server 2.x (the web version of VSCode) as an intrinsic service port.
  (#191, #194)
* BACKPORT/IMPROVE: Keep track of service process lifecycle expplicitly and allow service processes to
  be restarted if they terminate. (#183)
* FIX: Load sitecustomize.py that hooks user inputs only in the batch-mode execution.
  Previously it was loaded for all Python processes and prevented user prompts in pip and other tools.
  (#193)
* BACKPORT/FIX: Allow users to edit/override their dotfiles when spawning new containers. (#192)
* FIX: Image list update regression due to typo in the codes (#188)

19.09.17 (2020-03-08)
---------------------

* IMPROVE/FIX: Rewrite container lifecycle management using queues and persistent states (#187)
* MAINTENANCE: Update backend.ai-common dependency.

19.09.16 (2020-02-27)
---------------------

* MAINTENANCE: Update dependency packages.

19.09.15 (2020-02-21)
---------------------

* HOTFIX: Remove duplicate code block causing agent startup crashes, which is originated from
  a merge error.

19.09.14 (2020-02-17)
---------------------

* FIX: Corruption of resource allocation maps due to abnormal termination of containers and/or
  race conditions of resource cleanup handlers. (#180)
* FIX: .bashrc not loaded by the tmux session which is default-enabled in the ttyd intrinsic app.

19.09.13 (2020-02-10)
---------------------

* NEW: ttyd terminals now use a shared tmux session by default, making the container's shell session
  persistent across browser refreshes and intermittent connection failures, and also allowing
  pair-programming by sharing the ttyd session. (#168, #178)

19.09.12 (2020-02-10)
---------------------

* NEW: Add support for user-specific dotfiles population in session containers (#166)
* MAINTENANCE: Revamp the CI/CD pipelines (#173)

19.09.11 (2020-01-20)
---------------------

* FIX: Package version conflicts due to aiobotocore/botocore and python-dateutil, by removing no longer
  used codes and aiobotocore dependency.

19.09.10 (2020-01-09)
---------------------

* FIX: Support host-to-container PID mapping in older Linux kernels (lower than 4.1) which does not
  provide NSPid field in /proc task status.

sorna/agent/kernel.py
import asyncio
import codecs
from collections import OrderedDict
import enum
import io
import logging
import time

from async_timeout import timeout
import aiozmq
from namedlist import namedlist
import simplejson as json
import zmq

from sorna.common.utils import StringSetFlag

log = logging.getLogger(__name__)

# msg types visible to the API client.
# (excluding control signals such as 'finished' and 'waiting-input'
# since they are passed as separate status field.)
outgoing_msg_types = {'stdout', 'stderr', 'media', 'html', 'log', 'completion'}


class KernelFeatures(StringSetFlag):
    UID_MATCH = 'uid-match'
    USER_INPUT = 'user-input'
    BATCH_MODE = 'batch'
    QUERY_MODE = 'query'
    TTY_MODE = 'tty'


class ClientFeatures(StringSetFlag):
    INPUT = 'input'
    CONTINUATION = 'continuation'


class ExecutionPhase(enum.Enum):
    BUILD = 1
    RUN = 2


class ExpectedInput(enum.Enum):
    USER_CODE = 0
    USER_INPUT = 1
    CONTINUATION = 2


class InputRequestPending(Exception):
    def __init__(self, opts):
        super().__init__()
        self.opts = opts


class BuildFinished(Exception):
    pass


class UserCodeFinished(Exception):
    def __init__(self, opts):
        super().__init__()
        self.opts = opts


class ExecTimeout(Exception):
    pass


ResultRecord = namedlist('ResultRecord', [
    ('msg_type', None),
    ('data', None),
])


class KernelRunner:

    def __init__(self, kernel_id,
                 kernel_ip, repl_in_port, repl_out_port,
                 exec_timeout, client_features=None):
        self.started_at = None
        self.finished_at = None
        self.kernel_id = kernel_id
        self.mode = None
        self.phase: ExecutionPhase
        self.expects: ExepctedInput = ExpectedInput.USER_CODE
        self.kernel_ip = kernel_ip
        self.repl_in_port  = repl_in_port
        self.repl_out_port = repl_out_port
        self.input_stream = None
        self.output_stream = None
        assert exec_timeout >= 0
        self.exec_timeout = exec_timeout
        self.max_record_size = 10485760  # 10 MBytes
        self.completion_queue = asyncio.Queue(maxsize=128)
        self.read_task = None
        self.client_features = client_features or set()

        self.output_queue = None
        self.pending_queues = OrderedDict()
        self.current_run_id = None

    async def start(self):
        self.started_at = time.monotonic()

        self.input_stream = await aiozmq.create_zmq_stream(
            zmq.PUSH, connect=f'tcp://{self.kernel_ip}:{self.repl_in_port}')
        self.input_stream.transport.setsockopt(zmq.LINGER, 50)

        self.output_stream = await aiozmq.create_zmq_stream(
            zmq.PULL, connect=f'tcp://{self.kernel_ip}:{self.repl_out_port}')
        self.output_stream.transport.setsockopt(zmq.LINGER, 50)

        self.read_task = asyncio.ensure_future(self.read_output())
        has_continuation = ClientFeatures.CONTINUATION in self.client_features
        self.flush_timeout = 2.0 if has_continuation else None
        if self.exec_timeout > 0:
            self.watchdog_task = asyncio.ensure_future(self.watchdog())
        else:
            self.watchdog_task = None

    async def close(self):
        if self.read_task and not self.read_task.done():
            self.read_task.cancel()
            await self.read_task
        if self.watchdog_task and not self.watchdog_task.done():
            self.watchdog_task.cancel()
            await self.watchdog_task
        if (self.input_stream and
            not self.input_stream.at_closing() and
            self.input_stream.transport):
            # only when really closable...
            self.input_stream.close()
        if (self.output_stream and
            not self.output_stream.at_closing() and
            # only when really closable...
            self.output_stream.transport):
            self.output_stream.close()

    async def feed_batch_opts(self, opts):
        self.input_stream.write([
            b'build',
            opts.get('build', '').encode('utf8'),
        ])
        self.input_stream.write([
            b'exec',
            opts.get('exec', '').encode('utf8'),
        ])
        await self.input_stream.drain()

    async def feed_code(self, text):
        self.input_stream.write([b'code', text.encode('utf8')])

    async def feed_input(self, text):
        self.input_stream.write([b'input', text.encode('utf8')])

    async def feed_interrupt(self):
        self.input_stream.write([b'interrupt', b''])

    async def feed_and_get_completion(self, code_text, opts):
        payload = {
            'code': code_text,
        }
        payload.update(opts)
        self.input_stream.write([
            b'complete',
            json.dumps(payload).encode('utf8'),
        ])
        try:
            result = await self.completion_queue.get()
            self.completion_queue.task_done()
            return json.loads(result)
        except asyncio.CancelledError:
            return []

    async def watchdog(self):
        try:
            await asyncio.sleep(self.exec_timeout)
            if self.output_queue is not None:
                # TODO: what to do if None?
                self.output_queue.put(ResultRecord('exec-timeout', None))
        except asyncio.CancelledError:
            pass

    @staticmethod
    def aggregate_console(result, records, api_ver):

        if api_ver == 1:

            stdout_items = []
            stderr_items = []
            media_items = []
            html_items = []

            for rec in records:
                if rec.msg_type == 'stdout':
                    stdout_items.append(rec.data)
                elif rec.msg_type == 'stderr':
                    stderr_items.append(rec.data)
                elif rec.msg_type == 'media':
                    o = json.loads(rec.data)
                    media_items.append((o['type'], o['data']))
                elif rec.msg_type == 'html':
                    html_items.append(rec.data)

            result['stdout'] = ''.join(stdout_items)
            result['stderr'] = ''.join(stderr_items)
            result['media'] = media_items
            result['html'] = html_items

        elif api_ver == 2:

            console_items = []
            completions = []
            last_stdout = io.StringIO()
            last_stderr = io.StringIO()

            for rec in records:

                if last_stdout.tell() and rec.msg_type != 'stdout':
                    console_items.append(('stdout', last_stdout.getvalue()))
                    last_stdout.seek(0)
                    last_stdout.truncate(0)
                if last_stderr.tell() and rec.msg_type != 'stderr':
                    console_items.append(('stderr', last_stderr.getvalue()))
                    last_stderr.seek(0)
                    last_stderr.truncate(0)

                if rec.msg_type == 'completion':
                    completions.extend(json.loads(rec.data))
                elif rec.msg_type == 'stdout':
                    last_stdout.write(rec.data)
                elif rec.msg_type == 'stderr':
                    last_stderr.write(rec.data)
                elif rec.msg_type == 'media':
                    o = json.loads(rec.data)
                    console_items.append((rec.msg_type, (o['type'], o['data'])))
                elif rec.msg_type in outgoing_msg_types:
                    console_items.append((rec.msg_type, rec.data))

            if last_stdout.tell():
                console_items.append(('stdout', last_stdout.getvalue()))
            if last_stderr.tell():
                console_items.append(('stderr', last_stderr.getvalue()))

            result['console'] = console_items
            result['completions'] = completions
            last_stdout.close()
            last_stderr.close()

        else:
            raise AssertionError('Unrecognized API version')

    async def get_next_result(self, api_ver=2):
        # Context: per API request
        try:
            records = []
            with timeout(self.flush_timeout):
                while True:
                    rec = await self.output_queue.get()
                    if rec.msg_type in outgoing_msg_types:
                        records.append(rec)
                    self.output_queue.task_done()
                    if rec.msg_type == 'finished':
                        o = json.loads(rec.data) if rec.data else {}
                        raise UserCodeFinished(o)
                    elif rec.msg_type == 'build-finished':
                        raise BuildFinished
                    elif rec.msg_type == 'waiting-input':
                        o = json.loads(rec.data) if rec.data else {}
                        raise InputRequestPending(o)
                    elif rec.msg_type == 'exec-timeout':
                        raise ExecTimeout
        except asyncio.TimeoutError:
            result = {
                'status': 'continued',
            }
            type(self).aggregate_console(result, records, api_ver)
            self.resume_output_queue()
            self.expects = ExpectedInput.CONTINUATION
            return result
        except BuildFinished as e:
            self.phase = ExecutionPhase.RUN
            result = {
                'status': 'build-finished',
            }
            type(self).aggregate_console(result, records, api_ver)
            self.resume_output_queue()
            return result
        except UserCodeFinished as e:
            result = {
                'status': 'finished',
                'options': e.opts,
            }
            type(self).aggregate_console(result, records, api_ver)
            self.next_output_queue()
            self.expects = ExpectedInput.USER_CODE
            return result
        except ExecTimeout:
            result = {
                'status': 'exec-timeout',
            }
            log.warning('Execution timeout detected on kernel '
                        f'{self.kernel_id}')
            type(self).aggregate_console(result, records, api_ver)
            self.next_output_queue()
            self.expects = ExpectedInput.USER_CODE
            return result
        except InputRequestPending as e:
            result = {
                'status': 'waiting-input',
                'options': e.opts,
            }
            type(self).aggregate_console(result, records, api_ver)
            self.resume_output_queue()
            self.expects = ExpectedInput.USER_INPUT
            return result
        except asyncio.CancelledError:
            self.resume_output_queue()
            raise
        except:
            log.exception('unexpected error')
            raise

    async def attach_output_queue(self, run_id):
        # Context: per API request
        if run_id not in self.pending_queues:
            q = asyncio.Queue(maxsize=4096)
            activated = asyncio.Event()
            self.pending_queues[run_id] = (activated, q)
        else:
            activated, q = self.pending_queues[run_id]
        if self.output_queue is None:
            self.output_queue = q
        else:
            if self.current_run_id == run_id:
                # No need to wait if we are continuing.
                pass
            else:
                # If there is an outstanding ongoning execution,
                # wait until it has "finished".
                await activated.wait()
                activated.clear()
        self.current_run_id = run_id
        assert self.output_queue is q

    def resume_output_queue(self):
        '''
        Use this to conclude get_next_result() when the execution should be
        continued from the client.

        At that time, we need to reuse the current run ID and its output queue.
        We don't change self.output_queue here so that we can continue to read
        outputs while the client sends the continuation request.
        '''
        self.pending_queues.move_to_end(self.current_run_id, last=False)

    def next_output_queue(self):
        '''
        Use this to conclude get_next_result() when we have finished a "run".
        '''
        assert self.current_run_id is not None
        del self.pending_queues[self.current_run_id]
        self.current_run_id = None
        if self.pending_queues:
            # Make the next waiting API request handler to proceed.
            activated, q = self.pending_queues.popitem(last=False)
            self.output_queue = q
            activated.set()
        else:
            # If there is no pending request, just ignore all outputs
            # from the kernel.
            self.output_queue = None

    async def read_output(self):
        # We should use incremental decoder because some kernels may
        # send us incomplete UTF-8 byte sequences (e.g., Julia).
        decoders = (
            codecs.getincrementaldecoder('utf8')(),
            codecs.getincrementaldecoder('utf8')(),
        )
        while True:
            try:
                msg_type, msg_data = await self.output_stream.read()
                # TODO: test if save-load runner states is working
                #       by printing received messages here
                if len(msg_data) > self.max_record_size:
                    msg_data = msg_data[:self.max_record_size]
                log.warning(f'{msg_type!r} {msg_data!r}')
                try:
                    if msg_type == b'completion':
                        # As completion is processed asynchronously
                        # to the main code execution, we directly
                        # put the result into a separate queue.
                        await self.completion_queue.put(msg_data)
                    elif msg_type == b'stdout':
                        if self.output_queue is None:
                            continue
                        await self.output_queue.put(
                            ResultRecord(
                                'stdout',
                                decoders[0].decode(msg_data),
                            ))
                    elif msg_type == b'stderr':
                        if self.output_queue is None:
                            continue
                        await self.output_queue.put(
                            ResultRecord(
                                'stderr',
                                decoders[1].decode(msg_data),
                            ))
                    else:
                        # Normal outputs should go to the current
                        # output queue.
                        if self.output_queue is None:
                            continue
                        await self.output_queue.put(
                            ResultRecord(
                                msg_type.decode('ascii'),
                                msg_data.decode('utf8'),
                            ))
                except asyncio.QueueFull:
                    pass
                if msg_type == b'build-finished':
                    # finalize incremental decoder
                    decoders[0].decode(b'', True)
                    decoders[1].decode(b'', True)
                elif msg_type == b'finished':
                    # finalize incremental decoder
                    decoders[0].decode(b'', True)
                    decoders[1].decode(b'', True)
                    self.finished_at = time.monotonic()
            except (asyncio.CancelledError, aiozmq.ZmqStreamClosed, GeneratorExit):
                break
            except:
                log.exception('unexpected error')
                break
